# PES University, Bangalore
Established under Karnataka Act No. 16 of 2013

UE20CS312 - Data Analytics - Worksheet 5
Course instructor: Gowri Srinivasa, Professor Dept. of CSE, PES University

Designed by Yashas Kadambi, Dept. of CSE - yashasks@pesu.pes.edu



# NAME : Adarsh Subhas Nayak
# SRN:PES1UG20CS620

# Markov Chains and AB Testing

### Prerequisites
- Revise the following concepts
    - Markov Chains
    - Markov Chains with Absorbing states
    - Calculation of eventual probability of aborbtion
    - A/B Testing
- Install the following software
    - pandas
    - numpy

### Points
The problems in this worksheet are for a total of 10 points with each problem having a different weightage.
- Problem 1: 2 points
- Problem 2: 4 points
- Problem 3: 4 points

# Scenario 2

Its a freezing day in New york, Commisioner Wench has sent a report to Captain Holt that the 99th precinct has much lower reported crimes compared to other precincts. Upon Analysis Captain Holt decides to add feedback unit along with the 4 major units to analyse this descripency. All the units are mentioned below

1.   Major Crimes
2.   Traffic
3.   General crimes
4.   Feedback
5.   Theft

---
---
<br>

The initial probablity of a person going to a particular unit on a particular day is given as follows


<br>

Major Crimes | Traffic | General crimes | Feedback | Theft
:----------: | :-----: | :------------: | :--------: | :---:
0.3 | 0.4 | 0.1 | 0.15 | 0.05 
<br>


To measure how many people will go to the feedback unit, the personel files of all employees are give to the **_Move-o-Tron 99_** and it gives us the following matrix which shows us the probability of people moving from one unit to another on a particular day . It is known that the **_Move-o-Tron 99_** alwasy outputs matices which follow a first order Markov chain. 

| |Major Crimes|Traffic|General crimes|Feedback|Theft|
|---|---|---|---|---|---|
|Major Crimes|0\.002|0\.666|0\.31|0\.0|0\.022|
|Traffic|0\.466|0\.102|0\.222|0\.0|0\.21|
|General crimes|0\.022|0\.11|0\.502|0\.0|0\.366|
|Feedback|0\.0|0\.0|0\.0|1\.0|0\.0|
|Theft|0\.11|0\.122|0\.066|0\.0|0\.702|

As the people of New York are smart the will learn where all the units are present and hence the next days (day 1) distribution will be the distribution present at the end of the current day (day 0). Captain holt want to check if the matrix given by the **_Move-o-Tron_** can be used to model the footfall.



## Problem 1 (2 points)

1. What technique can be used to model the probability of people going to the correct unit to report thier crime after N days? (0 points)
2. Is the chain irreducible? Justify (0.5 point)
3. What will be the intital probability of a person going to a particular unit after 1 day, 2 days, 10 days, 1000 days and 1001 days. (1 point)
 
 Hint: Use the  Chapman−Kolmogorov relationship

     ```python
    # C = A.B
    matrix_C = np.dot(matrix_A, matrix_B)

    # C = A.(B^4) can be replaced by
    matrix_C = matrix_A
    for _ in range(4):
        matrix_C = np.dot(matrix_C, matrix_B)
    ```

4. What can you say about the markov chain from state of intital probability of a person going to a particular unit after 1000 and 1001 days? (0.5 points)

In [1]:
# Importing Libraries
%matplotlib inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt


In [2]:
# encoding the probabilities as a numpy array
trans_array = np.array([
    [0.002, 0.666, 0.31, 0, 0.022],
    [0.466, 0.102, 0.222, 0, 0.21],
    [0.022, 0.11, 0.502, 0, 0.366],
    [0, 0, 0, 1, 0],
    [0.11, 0.122, 0.066, 0, 0.702]
])
# ensures that the probabilities sum to 1
assert(trans_array[0].sum() == 1.0)
assert(trans_array[1].sum() == 1.0)
assert(trans_array[2].sum() == 1.0)
assert(trans_array[3].sum() == 1.0)
assert(trans_array[4].sum() == 1.0)



# encoding the initial probability of as a numpy array
state = np.array([[0.3, 0.4, 0.1, 0.15, 0.05]])
assert(state[0].sum() == 1.0)

In [8]:

print(" 01 :  MARKOV CHAIN ")

print(" 02:   No, Its reducable if and only if there are two or more communication classes.")

print("3:")
state_one=np.dot(state, trans_array)
print(f' state after 1day : {state_one.squeeze()}')
#After 2 day
state_two=np.dot(state_one, trans_array)
print(f' state after 1day : {state_two.squeeze()}')
#After 10 days
state_10 = state
for _ in range(10):
  state_10 = np.dot(state_10, trans_array)
print(f' state after 10days : {state_10.squeeze()}')
#After 1000days
state_1000 = state
for _ in range(1000):
  state_1000 = np.dot(state_1000, trans_array)
print(f' state after 1000days : {state_1000.squeeze()}')
#After 1001 days
state_1001=np.dot(state_1000,trans_array)
print(f' state after 1001days : {state_1001.squeeze()}')

print('QUESTION 04')
print('After looking at the states after 1000 days and 1001 days ... one can notice zero difference/change in probability distribution.So therefore represents Stationary probability distribution!!')

QUESTION 01 :  MARKOV CHAIN 
Question 02:   No, Its reducable if and only if there are two or more communication classes.
QUESTION 03
 state after 1days : [0.1947 0.2577 0.2353 0.15   0.1623]
 state after 1days : [0.1435072 0.2016392 0.2463988 0.15      0.2584548]
 state after 10days : [0.12139467 0.16406836 0.1973962  0.15       0.36714077]
 state after 1000days : [0.1214373  0.16411091 0.19739717 0.15       0.36705462]
 state after 1001days : [0.1214373  0.16411091 0.19739717 0.15       0.36705462]
QUESTION 04
After looking at the states after 1000 days and 1001 days ... one can notice zero difference/change in probability distribution.So therefore represents Stationary probability distribution!!


---

After analysing the model Captain holt calls the squad and educates them to ask people to give feedbacks. And the details of the squad are given to the **_Move-o-Tron 99_**. After reanalyising the report the **_Move-o-Tron 99_** gave out a new Matrix, which is shown below

| |Major Crimes|Traffic|General crimes|Feedback|Theft|
|---|---|---|---|---|---|
|Major Crimes|0\.002|0\.666|0\.01|0\.02|0\.302|
|Traffic|0\.466|0\.102|0\.02|0\.032|0\.38|
|General crimes|0\.0|0\.0|1\.0|0\.0|0\.0|
|Feedback|0\.0|0\.0|0\.0|1\.0|0\.0|
|Theft|0\.11|0\.122|0\.066|0\.004|0\.698|

<br>
Considering the new report the model has to be re-evaluated. The initial probablity of a person going to a particular unit on a particular day remains the same.
<br>





## Problem 2 (4 points)


1. Is the chain irreducible? Justify (0.5 point)
2. What will be the intital probability of a person going to a particular unit after 1 day, 2 days, 10 days, 1000 days and 1001 days. (1 point)
 
 Hint: Use the  Chapman−Kolmogorov relationship

3. What can you say about the markov chain from state of intital probability of a person going to a particular unit after 1000 and 1001 days? (0.5 points)


4. Summer Edgecombe is Confidential Informant (CI) to the Officer Kimbal Cho and comes in every day to the police station. If on the first day she goes to the Major crimes Unit what will be the probability that she gives a feedback? (2 points)

    ```python
    # np.delete()
    # https://note.nkmk.me/en/python-numpy-delete/#:~:text=Using%20the%20NumPy%20function%20np,from%20the%20NumPy%20array%20ndarray%20.&text=Specify%20the%20axis%20(dimension)%20and,a%20slice%20or%20a%20list.

    print(a)
    # [[ 0  1  2  3]
    #  [ 4  5  6  7]
    #  [ 8  9 10 11]]

    print(np.delete(a, 1, 0))
    # [[ 0  1  2  3]
    #  [ 8  9 10 11]]

    print(np.delete(a, 1, 1))
    # [[ 0  2  3]
    #  [ 4  6  7]
    #  [ 8 10 11]]

    # Deleting multiple rows or columns
    print(np.delete(a, [0, 3], 1))
    # [[ 1  2]
    #  [ 5  6]
    #  [ 9 10]]

    # Deleting rows and columns
    print(np.delete(np.delete(a, 1, 0), 1, 1))
    # [[ 0  2  3]
    #  [ 8 10 11]]

    # matrix multiplication or cross pdt C = A*B
    matrix_C = matrix_A @ matrix_B # matrix_C = np.matmul(matrix_A, matrix_B)

    ```

In [3]:
# encoding the probabilities as a numpy array
trans_array = np.array([
    [0.002, 0.666, 0.01, 0.020, 0.302],
    [0.466, 0.102, 0.02, 0.032, 0.38],
    [0.0, 0.0, 1, 0.0, 0.0],
    [0, 0, 0, 1, 0],
    [0.11, 0.122, 0.066, 0.004, 0.698]
])

# ensures that the probabilities sum to 1
assert(trans_array[0].sum() == 1.0)
assert(trans_array[1].sum() == 1.0)
assert(trans_array[2].sum() == 1.0)
assert(trans_array[3].sum() == 1.0)
assert(trans_array[4].sum() == 1.0)

# encoding the initial probability of as a numpy array
state = np.array([[0.3, 0.4, 0.1, 0.15, 0.05]])
assert(state[0].sum() == 1.0)


In [19]:
print('01 :')
print('no its not reducable as feedbak unit ot general crimes unit isnt accessable.')
print('02 :')
state_one=np.dot(state, trans_array)
print(f' state after 1day : {state_one.squeeze()}')
#After 2 day
state_two=np.dot(state_one, trans_array)
print(f' state after 1day : {state_two.squeeze()}')
#After 10 days
state_10 = state
for _ in range(10):
  state_10 = np.dot(state_10, trans_array)
print(f' state after 10days : {state_10.squeeze()}')
#After 1000days
state_1000 = state
for _ in range(1000):
  state_1000 = np.dot(state_1000, trans_array)
print(f' state after 1000days : {state_1000.squeeze()}')
#After 1001 days
state_1001=np.dot(state_1000,trans_array)
print(f' state after 1001days : {state_1001.squeeze()}')

print('QUESTION 03')
print('Now we notice some change in probability distribution in states after 1000days and 1001days')

print('QUESTION 04')
Q=np.delete(np.delete(trans_array,[2,3],0),[2,3],1)
print(f'MATRIX Q:\n{Q[0]}\n{Q[1]}\n{Q[2]}')
R=np.delete(np.delete(trans_array,[2,3],0),[0,1,4],1)
print(f'MATRIX R:\n{R[0]}\n{R[1]}\n{R[2]}')
I=np.eye(3)
print(f'MATRIX I:\n{I[0]}\n{I[1]}\n{I[2]}')
F= np.linalg.inv(I-Q)
print(f'MATRIX F:\n{F[0]}\n{F[1]}\n{F[2]}')
FR= F @ R
print(f'MATRIX FR:\n{FR[0]}\n{FR[1]}\n{FR[2]}')

print(f'Probabilty: 0.25455')

QUESTION 01
no its not reducable as feedbak unit ot general crimes unit isnt accessable.
QUESTION 02
 state after 1days : [0.1925 0.2467 0.1143 0.169  0.2775]
 state after 1days : [0.1458722 0.1872234 0.139474  0.1818544 0.345576 ]
 state after 10days : [0.07693332 0.09688565 0.3362719  0.24330485 0.24660428]
 state after 1000days : [8.97378472e-28 1.13004326e-27 6.60595331e-01 3.39404669e-01
 2.87688168e-27]
 state after 1001days : [8.44851901e-28 1.06389804e-27 6.60595331e-01 3.39404669e-01
 2.70848815e-27]
QUESTION 03
Now we notice some change in probability distribution in states after 1000days and 1001days
QUESTION 04
MATRIX Q:
[0.002 0.666 0.302]
[0.466 0.102 0.38 ]
[0.11  0.122 0.698]
MATRIX R:
[0.01 0.02]
[0.02  0.032]
[0.066 0.004]
MATRIX I:
[1. 0. 0.]
[0. 1. 0.]
[0. 0. 1.]
MATRIX F:
[4.0279365  4.26333958 9.39240352]
[3.27006043 4.80437252 9.31529737]
[ 2.78814698  3.49371126 10.49546579]
MATRIX FR:
[0.74544479 0.25455521]
[0.74359768 0.25640232]
[0.79045644 0.20954356]


##Problem 3 (4 points)

It seems that there is a bug in **_Move-o-Tron 99_** which makes general crimes and feedback units as absorbing states. After updating the software of **_Move-o-Tron 99_**, Captain Holt wants to find out the effect that Amy Santiago has on the probability of a person giving feedback. So one matrix is generated including Santiago and the other one without. 

Matrix 1 (With Santiago)

| |Major Crimes|Traffic|General crimes|Feedback|Theft|
|---|---|---|---|---|---|
|Major Crimes|0\.002|0\.232|0\.31|0\.434|0\.022|
|Traffic|0\.426|0\.102|0\.222|0\.04|0\.21|
|General crimes|0\.03|0\.11|0\.2|0\.294|0\.366|
|Feedback|0\.003|0\.176|0\.321|0\.3|0\.2|
|Theft|0\.11|0\.422|0\.166|0\.1|0\.202|

Matrix 2 (Without Santiago)

| |Major Crimes|Traffic|General crimes|Feedback|Theft|
|---|---|---|---|---|---|
|Major Crimes|0\.11|0\.222|0\.092|0\.374|0\.202|
|Traffic|0\.03|0\.11|0\.2|0\.294|0\.366|
|General crimes|0\.002|0\.232|0\.31|0\.434|0\.022|
|Feedback|0\.466|0\.102|0\.02|0\.032|0\.38|
|Theft|0\.003|0\.176|0\.321|0\.3|0\.2|

<br>

1. How can you find out the effect that Santiago has on the probability of feedback? (1 point)

2. What effect does Santiago have one the probability of getting feedback? (1 point)

    Note: The initial probablity of a person going to a particular unit on a particular day remains the same

3. Name the test Captain Holt is performing. (0.5 points)

Lina Ginetti reports to Captain Holt that the there two kinds of days in the precicnt _"There are normal days and then there are days where workflow is dismal with a tiny dash of pathetic."_. Captain Holt decided to sample the initial probablity of a person going to a particular unit on a good day and a bad day.

4. Without the information about these inital probabilities, can you tell if there is any difference in the probability of getting a feedback? Explain. (1.5 points)



In [4]:
# encoding the probabilities as a numpy array
# With Santiago
trans_array_with_amy = np.array([
    [0.002, 0.232, 0.31, 0.434, 0.022],
    [0.426, 0.102, 0.222, 0.04, 0.21],
    [0.03, 0.11, 0.20, 0.294, 0.366],
    [0.003, 0.176, 0.321, 0.3, 0.2],
    [0.11, 0.422, 0.166, 0.1, 0.202]
])

# Without Santiago
trans_array_without_amy = np.array([
    [0.11, 0.222, 0.092, 0.374, 0.202],
    [0.03, 0.11, 0.20, 0.294, 0.366],
    [0.002, 0.232, 0.31, 0.434, 0.022],
    [0.466, 0.102, 0.02, 0.032, 0.38],
    [0.003, 0.176, 0.321, 0.3, 0.2]
])

# ensures that the probabilities sum to 1
assert(trans_array_with_amy[0].sum() == 1.0)
assert(trans_array_with_amy[1].sum() == 1.0)
assert(trans_array_with_amy[2].sum() == 1.0)
assert(trans_array_with_amy[3].sum() == 1.0)
assert(trans_array_with_amy[4].sum() == 1.0)

assert(trans_array_without_amy[0].sum() == 1.0)
assert(trans_array_without_amy[1].sum() == 1.0)
assert(trans_array_without_amy[2].sum() == 1.0)
assert(trans_array_without_amy[3].sum() == 1.0)
assert(trans_array_without_amy[4].sum() == 1.0)

# encoding the initial probability of as a numpy array
state = np.array([[0.3, 0.4, 0.1, 0.15, 0.05]])
assert(state[0].sum() == 1.0)


In [32]:
print('QUESTION 01')
print(f'Find the Stationary Distribution of each of the transition matrix and then comparing the corresponding probability')
print('QUESTION 02')

print('Varient with including Amy Santiango')
state_1000_with_amy =state
for _ in range(1000):
  state_1000_with_amy = np.dot(state_1000_with_amy , trans_array_with_amy)
print(f'\t state after 1000days = {state_1000_with_amy.squeeze()}')
state_1001_with_amy = np.dot(state_1000_with_amy , trans_array_with_amy)
print(f'\t state after 1001days = {state_1001_with_amy.squeeze()}')


print('Varient without including Amy Santiango')

state_1000_without_amy =state
for _ in range(1000):
  state_1000_without_amy = np.dot(state_1000_without_amy , trans_array_without_amy)
print(f'\t state after 1000days = {state_1000_without_amy.squeeze()}')
state_1001_without_amy = np.dot(state_1000_without_amy , trans_array_without_amy)
print(f'\t state after 1001days = {state_1001_without_amy.squeeze()}')



print(f'{state_1000_with_amy.squeeze()[3]} is lesser than {state_1000_without_amy.squeeze()[3]} ')
print('Hence, probablity of getting a feedback with amy santiago working is lesser than without amy')

print('QUESTION 03')
print('AB TESTING')

print('QUESTION 04')
print('Yes,stationary distribution does not depend on the initial states. so there is no difference in the probability of getting a feedback')


QUESTION 01
Find the Stationary Distribution of each of the transition matrix and then comparing the corresponding probability
QUESTION 02
Varient with including Amy Santiango
	 state after 1000days = [0.12001975 0.20607988 0.23649985 0.21700347 0.22039705]
	 state after 1001days = [0.12001975 0.20607988 0.23649985 0.21700347 0.22039705]
Varient without including Amy Santiango
	 state after 1000days = [0.14495178 0.16283362 0.18658671 0.26400004 0.24162786]
	 state after 1001days = [0.14495178 0.16283362 0.18658671 0.26400004 0.24162786]
0.21700346817098676 is lesser than 0.2640000386809766 
Hence, probablity of getting a feedback with amy santiago working is lesser than without amy
QUESTION 03
AB TESTING
QUESTION 04
Yes,stationary distribution does not depend on the initial states. so there is no difference in the probability of getting a feedback
